zkvm/src/constraints.rs

```diff
@@ -89,21 +92,64 @@ impl Constraint {
         .map_err(|e| VMError::R1CSError(e))
     }
 
-    fn flatten<CS: r1cs::RandomizedConstraintSystem>(self, cs: &mut CS) -> r1cs::LinearCombination {
+    fn flatten<CS: r1cs::RandomizedConstraintSystem>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(r1cs::LinearCombination, Option<Scalar>), r1cs::R1CSError> {
         match self {
-            Constraint::Eq(expr1, expr2) => expr1.to_r1cs_lc() - expr2.to_r1cs_lc(),
+            Constraint::Eq(expr1, expr2) => {
+                let lc = expr1.to_r1cs_lc() - expr2.to_r1cs_lc();
+                match (expr1.eval(), expr2.eval()) {
+                    (Some(x), Some(y)) => Ok((lc, Some(x.to_scalar() - y.to_scalar()))),
+                    (_, _) => Ok((lc, None)),
+                }
+            }
             Constraint::And(c1, c2) => {
-                let a = c1.flatten(cs);
-                let b = c2.flatten(cs);
+                let (a, a_assg) = c1.flatten(cs)?;
+                let (b, b_assg) = c2.flatten(cs)?;
                 let z = cs.challenge_scalar(b"ZkVM.verify.and-challenge");
-                a + z * b
+                match (a_assg, b_assg) {
+                    (Some(a_assg), Some(b_assg)) => Ok((a + z * b, Some(a_assg + z * b_assg))),
+                    (_, _) => Ok((a + z * b, None)),
+                }
             }
             Constraint::Or(c1, c2) => {
-                let a = c1.flatten(cs);
-                let b = c2.flatten(cs);
+                let (a, a_assg) = c1.flatten(cs)?;
+                let (b, b_assg) = c2.flatten(cs)?;
                 // output expression: a * b
                 let (_l, _r, o) = cs.multiply(a, b);
-                r1cs::LinearCombination::from(o)
+                match (a_assg, b_assg) {
+                    (Some(a_assg), Some(b_assg)) => {
+                        Ok((r1cs::LinearCombination::from(o), Some(a_assg * b_assg)))
+                    }
+                    (_, _) => Ok((r1cs::LinearCombination::from(o), None)),
+                }
+            }
+            Constraint::Not(c1) => {
+                let (x_lc, x_assg) = c1.flatten(cs)?;
+                let (xy, xw, y_assg) = match x_assg {
+                    Some(x_assg) => {
+                        let is_zero = x_assg.ct_eq(&Scalar::zero());
+                        let y_assg = Scalar::conditional_select(
+                            &Scalar::zero(),
+                            &Scalar::one(),
+                            is_zero.into(),
+                        );
+                        let w = Scalar::conditional_select(&x_assg, &Scalar::one(), is_zero.into());
```

Comment:
```suggestion
                        let w = Scalar::conditional_select(&x_assg, &Scalar::one(), is_zero);
```

`is_zero` is already a `Choice`.

Reply:
```suggestion
                        let w = Scalar::conditional_select(&Scalar::one(), &x_assg, is_zero);
```

Does `choice==true` select the first or the second variant?

Reply:
if x==0, we need w=1 to make it invertible

Reply:
hm, I see inconsistent documentation — the function doc says that it chooses `a` if `Choice(0)` but then the subtle docs online say the opposite. will investigate

Reply:
based on https://doc.dalek.rs/src/subtle/lib.rs.html#364 I think that `conditional_select` chooses `a` on `Choice(0)`, so `a` on `false` and `b` on `true`

Reply:
omg